doc/design/block.md

```diff
@@ -0,0 +1,210 @@
+
+
+```c++
+// Information in SymbolTable is enough to trace the dependency graph. So maybe
+// the Eval() interface takes a SymbolTable is enough.
+class SymbolTable {
+ public:
+  SymbolTable(SymbolTable* parent) : parent_(parent) {}
+
+  OpDesc* NewOp(const string& name="");
+
+  // TODO determine whether name is generated by python or C++
+  // currently assume that a unique name will be generated by C++ if the
+  // argument name left default.
+  VarDesc* NewVar(const string& name="");
+
+  // find a VarDesc by name, if recursive true, find parent's SymbolTable
+  // recursively.
+  // this interface is introduced to support InferShape, find protobuf messages
+  // of variables and operators, pass pointers into InferShape.
+  // operator
+  //
+  // NOTE maybe some C++ classes such as VarDescBuilder and OpDescBuilder should
+  // be proposed and embedded into pybind to enable python operate on C++ pointers.
+  VarDesc* FindVar(const string& name, bool recursive=true);
+
+  OpDesc* FindOp(const string& name);
+
+  BlockDesc Serialize() const;
+  void DeSerialize();
+
+ private:
+  SymbolTable* parent_;
+
+  map<string, OpDesc> ops_;
+  map<string, VarDesc> vars_;
+};
+```
+
+After all the description of variables and operators is added into SymbolTable,
+the block has enough information to run.
+
+To make block simpler, we use another class `RuntimeTable` to create Variable and Operator objects. `RuntimeTable` is free to add memory optimizements in the future.
+
+The `Block` class takes a `BlockDesc` as input, and provide `Run` and `InferShape` functions.
+
+
+```c++
+namespace {
+// A RuntimeTable maintains all the runtime environment symbols of a block,
+// including the variables and operators.
+// NOTE the implementation of runtime environment is seperated from Block
+// to make it easier to improve memory usage and some other improvements
+// based on a block's description.
+// NOTE this name is aweful, and should be updated latter.
+class RuntimeTable {
+ public:
+  // scope: the scope to create new variables.
+  RuntimeTable(const BlockDesc& desc, const Scope& scope);
+
+  vector<OperatorBase*> ops() { return ops_; }
+
+  bool IsValid() const { return ops_.empty(); }
+
+ protected:
+  void CreateVars(const Scope& scope);
+  void CreateOps();
+
+ private:
+  SymbolTable symbol_tabel_;
+  vector<OperatorBase*> ops_;
+};
+}
+
+
+class Block : OperatorBase {
+public:
+  // desc may be serialized from a RuntimeTable or not.
+  Block(const BlockDesc& desc) desc_(desc) {}
+
+  void InferShape(const framework::Scope& scope) const override {
+    // should run InferShape first.
+    if (!runtime_table_.IsValid()) {
+      runtime_table_ = RuntimeTable(desc_, scope);
+    }
+    for (auto& op : runtime_table_.ops()) {
+      op->InferShape(scope);
+    }
+  }
+
+  void Run(const framework::Scope& scope,
+           const platform::DeviceContext& dev_ctx) const override {
```

Comment:
What happens if ops from a block require to run on different device contexts? (e.g., one OP can run on CPU only, other OPs must run on GPU)

Reply:
this is a historical issue, Block inherent from `OperatorBase`, so the same interface of `Run`.

Reply:
I think multi-device is a very important feature and we should take this into consideration at current desgin.
The basic elements of Block(or Graph) is actually Nodes and Edges. Node has a device attribute to decide which device to run. Device includes CPU/MKL/CUDA/FPGA/etc. Some developers from inf will contribute FPGA code for refactoring Paddle. And FPGA is not suitable for all operators, and will only optimize for specific operators. So, multi-device(CPU/FPGA) is needed.
Please refer to #3943 

And, What's the relationship between these two concepts, Block and Graph? If Block is Graph, and the private member of Block should be Nodes and Edges. The Graph mainly describes data dependency and control dependency. And Graph will run by an Executor, the executor will support multi-device executing.

Reply:
better to discuss this in another issue.